gbn/queue.go

```diff
@@ -114,28 +236,100 @@ func (q *queue) resend() error {
 	q.topMtx.RUnlock()
 
 	if base == top {
+		q.awaitingCatchUpMu.Unlock()
+
 		return nil
 	}
 
+	// Prepare the queue for awaiting the resend catch up.
+	q.awaitedACK = (q.cfg.s + top - 1) % q.cfg.s
+	q.awaitedNACK = top
+
+	q.cfg.log.Tracef("Set awaitedACK to %d & awaitedNACK to %d",
+		q.awaitedACK, q.awaitedNACK)
+
+	q.awaitingCatchUp = true
+
+	// Drain the caughtUpSignal channel, in case no proceedAfterTime
+	// func was executed after the last resend catch up.
+	select {
+	case <-q.caughtUpSignal:
+	default:
+	}
+
 	q.cfg.log.Tracef("Resending the queue")
 
 	for base != top {
 		packet := q.content[base]
 
 		if err := q.cfg.sendPkt(packet); err != nil {
+			q.awaitingCatchUpMu.Unlock()
+
 			return err
 		}
+
 		base = (base + 1) % q.cfg.s
 
 		q.cfg.log.Tracef("Resent %d", packet.Seq)
 	}
 
+	// We hold the awaitingCatchUpMu mutex for the duration of the resend to
+	// ensure that we don't process the delayed ACKs for the packets we are
+	// resending, during the resend. If that would happen, we would start
+	// the "proceedAfterTime" function timeout while still resending
+	// packets. That could mean that the NACK that the resent packets will
+	// trigger, might be received after the timeout has passed. That would
+	// cause the resend loop to trigger once more.
+	q.awaitingCatchUpMu.Unlock()
```

Comment:
Thanks, I've updated your syncer slightly to also be able to process the awaited ACKs + NACKs while we are actually resending the queue, and not block by holding the lock while actually resending the queue.
This new version can trigger the edge case that the NACK that the resent packets will trigger, might be received after the proceedAfterTime timeout has passed. But as this would only cause one extra resend loop in the rare case that this would happen, I think it's worth the tradeoff :)!